src/analyze/events.py

```diff
@@ -1,98 +1,136 @@
-all_recs = []
-with gzip.open(path, "rt") as gid:
-    all_recs = json.load(gid)
-
-# strptime for posix time stamps
-tfmt = "%Y-%m-%dT%H:%M:%S.%fZ"
+
+# coding: utf-8
+
+# # Visualizing Caliper Events:<br />An Example Using Python With `pandas` and `numpy`
+# > 💡This notebook requires a few Python modules in addition to the common ones.  See the ["Prerequisites" section](#Prerequisites) at the end of this notebook for more information, especially if you see error messages or the notebook doesn't run correctly.
+# 
+# ## Introduction
+# 
+# This notebook reads Caliper events that were generated by an application.  For each course, it graphs the number of events that were found for each hour of the day, day of the week, and day of the year.
+# 
+# It was written to work with Caliper v1.0 events generated by the Problem Roulette (v1.0) application.  However, it may work with other event data JSON that contains "`.eventTime`" and "`.group.name`" attributes.
+
+# ## Configuration
+# 
+# The values of the variables in the cell below may be changed to specify the locations of the input data and output visualizations and other options.
+
+# In[1]:
+
+
+# Path to input data file.  Files that have been compressed with gzip are supported.
+path = '../../tmp/pr.jsonl'
+
+# Is the input JSONL (JSON Lines)? (If False, input will be treated as plain JSON)
+# The JSONL format (AKA NDJSON, newline-delimited JSON) contains complete JSON objects
+# separated only by line breaks (newline or carriage return)
+jsonlInput = True
+
+# Output to PDF?  (If False, display output inline with Jupyter Notebook)
+pdfOutput = True
+
+# Name of PDF output file.  (Ignored if pdfOutput == False)
+pdfFilename = 'events.pdf'
+
+
+# ## Initialization
+# Load the various Python modules needed.
+
+# In[2]:
+
+
+from eventsTools import *
+
+
+# ## Load the JSON Data Input File
+
+# In[3]:
+
+
+all_recs = loadEventJson(path, jsonlInput)
+
+print('Loaded {} records.'.format(len(all_recs)))
+
+
+# ## Extract Data From Caliper Events
+
+# In[4]:
+
 
 # Extract some fields for each event; agg is a map from course names
 # to event time stamps, the event times are split into columns
 # according to the time.struct_time fields:
 # https://docs.python.org/3/library/time.html#time.struct_time
 agg = []
 tfields = ['tm_hour', 'tm_mday', 'tm_min', 'tm_mon', 'tm_wday', 'tm_yday', 'tm_year']
-for rec in all_recs:
-
-    # This seems to be where the good information is
-    js = json.loads(rec['raw'])
 
-    # Extract the user id/, but not currently used here
-    if 'actor' not in js or js['actor'] is None:
-        continue
-    actor = js['actor']
-    user_url = actor["@id"]
-    user = user_url.split(":")
-    if len(user) < 3:
+for js in all_recs:
+    user = extractEventActorUserId(js)
+    if (user is False) or (len(user) < 3):
         continue
 
-    # Extract the time
-    if 'group' not in js or js['group'] is None:
+    course = extractEventCourseName(js)
+    if course is False:
```

Comment:
Good question.  I believe the `is` keyword is considered more "Pythonic".  It can't be used with everything, but it's usually seen in tests as `is False`, `is None`, etc.  Also, `!course` evaluates to true when the value of `course` is `False`, `None`, and possibly when it's undefined.

"Pythonic" style isn't _always_ good, in my opinion.  This is one of those cases I like it, because it reads more like human language.

FWIW, Chris Brooks described "pandorable" in his course, but I think that's just the Pandas community's misappropriation of the "Pythonic" term.